Add new test: vfd_swmr_attrdset_writer.c

vchoi-hdfgroup opened the pull request.

vchoi-hdfgroup committed:
Merge pull request #1 from HDFGroup/feature/vfd_swmr

Feature/vfd swmr

vchoi-hdfgroup committed:
Merge pull request #2 from vchoi-hdfgroup/feature/vfd_swmr

Feature/vfd swmr

vchoi-hdfgroup committed:
Merge pull request #3 from HDFGroup/feature/vfd_swmr

Feature/vfd swmr

vchoi-hdfgroup committed:
Merge pull request #4 from vchoi-hdfgroup/feature/vfd_swmr

Feature/vfd swmr

Commit:
Add new test vfd_swmr_attrdset_writer.c.
It is to test attribute handling for different dataset types.

raylu-hdf commented on test/vfd_swmr_attrdset_writer.c:
```diff
@@ -0,0 +1,2017 @@
+ *      -- Remove sufficient attributes to allow deletion of object header continuation block
+ *      -- Transition from compact to dense attribute storage
+ *      -- Transition from dense to compact attribute storage
+ */
```
I don't think the three headers below are necessary.

vchoi-hdfgroup replied:
Removed as indicated.

vchoi-hdfgroup replied:
[Dana used the H5_HAVE_WIN32_API to skip the VFD SWMR tests on Windows, e.g. in vfd_swmr_group_writer.c. That's a recent change. I think you may do the same for this new test.]
Add that in as indicated.

kyang2014 replied:
I will check the current vfd_swmr_group_writer.c. If it is not there, I will add it in.

raylu-hdf reviewed (commented):
Dana used the `H5_HAVE_WIN32_API` to skip the VFD SWMR tests on Windows, e.g. in vfd_swmr_group_writer.c.  That's a recent change.  I think you may do the same for this new test.

Commit:
(1) Remove header files as indicated by PR review comment.
(2) Add ifdef for windows as indicated by PR review comment.
(3) Misc. cleanup.

vchoi-hdfgroup committed:
Merge pull request #5 from HDFGroup/feature/vfd_swmr

Feature/vfd swmr

vchoi-hdfgroup committed:
Merge pull request #6 from vchoi-hdfgroup/feature/vfd_swmr

Feature/vfd swmr

raylu-hdf commented on test/vfd_swmr_attrdset_writer.c:
```diff
@@ -0,0 +1,2015 @@
+        else
+            HDsprintf(vl_which, "%u %u %u %u %u", which, which+1, which+2, which+3, which+4);
+
+        read_vl_which = HDmalloc(sizeof("9999999999"));
```
This allocated memory needs to be freed.

kyang2014 replied:
Also, in the group test, to verify the vl str attribute,  a HDfree_memory() is called after the attribute value is read and  is compared. I copied it from static void test_write_vl_string_attribute(void) in tvlstr.c under /test.  However, the group test only handles one VL string which is like test_write_vl_string_attribute(). The VL in this test has several elements.

vchoi-hdfgroup replied:
Corrected.

raylu-hdf commented on test/vfd_swmr_attrdset_writer.c:
```diff
@@ -0,0 +1,2015 @@
+/* This will be removed when this routine is moved to vfd_swmr_common.c */
+/* Sleep for `tenths` tenths of a second */
+static void
+decisleep(uint32_t tenths)
```
I put the definition of this function in `vfd_swmr_common.c` in my pull request for the zoo test.  I already merged my PR.

vchoi-hdfgroup replied:
Corrected.

kyang2014 commented on test/vfd_swmr_attrdset_writer.c:
```diff
@@ -0,0 +1,2015 @@
+		"\n",
+		progname);
+	exit(EXIT_FAILURE);
+} /* usgae() */
```
typo: should be usage()

vchoi-hdfgroup replied:
Corrected.

kyang2014 commented on test/vfd_swmr_attrdset_writer.c:
```diff
@@ -0,0 +1,2015 @@
+            TEST_ERROR;
+        }
+
+        if(H5Pset_alloc_time(dcpl, H5D_ALLOC_TIME_EARLY) < 0) {
```
I think the early allocation is chosen for the testing purpose. Should we have a test for the default case, which I believe is more common. Or it doesn't matter. Note this is a question rather than a request to add such a test.

vchoi-hdfgroup replied:
My thought is: the zoo test already tested contiguous dataset with default allocation, so I use early allocation for the contiguous dataset here.

vchoi-hdfgroup replied:
[Add sufficient attributes to force creation of object header continuation block]
[Remove sufficient attributes to allow deletion of object header continuation block]

For the verification of object header continuation block, please see verify_storage_cont_real() where ninfo.hdr.nchunks is 1 and > 1 as attributes are added and storage is moving between compact and dense.

kyang2014 commented on test/vfd_swmr_attrdset_writer.c:
```diff
@@ -0,0 +1,2015 @@
+
+/*
+ * Perform the attribute action on the specified dataset.
+ */
```
Maybe add description for this function. The action, s, and did are straightforward. But the "step" may not be obvious.

vchoi-hdfgroup replied:
Added.

kyang2014 replied:
OK, good and make sense.

kyang2014 reviewed (approved):
May add more comments to explain the test procedure. Especially for the followings:
1)  -- Add sufficient attributes to force creation of object header continuation block
 *      -- Remove sufficient attributes to allow deletion of object header continuation block
How this is verified. Perhaps this is just to me.
2) If I understand correctly, the program doesn't limit the number of attributes to add with the parameter asteps. This is a good robust test to test the addition of many attributes for one dataset.

Commit:
Modifications based on PR feedback.
